base/src/Data/Macaw/AbsDomain/JumpBounds.hs

```diff
@@ -661,89 +661,154 @@ addRangePred cns v p =
+addUpperBoundPred ::
+  RegisterInfo (ArchReg arch) =>
+  BlockIntraStackConstraints arch ids -> NatRepr n -> Value arch ids (BVType n) -> Integer ->
+  Either String (BranchConstraints arch ids)
+addUpperBoundPred cns w x u
+  | u < 0 = Left "Upper bound excludes all possible values"
+  | otherwise = addRangePred cns (intraStackValueExpr cns x) $! mkUpperBound w (fromInteger u)
+
+clamp :: NatRepr w -> Integer -> Integer
+clamp w x
+  | x < 0 = 0
+  | x > maxUnsigned w = maxUnsigned w
+  | otherwise = x
+
+-- | @addWithinRangePred cns w l x u@ adds a @l <= x <= u@ constraint over the `w`-bit-truncated values.
+-- Fails when the bound excludes all possible values.
+-- Note: The arguments are ordered so that it reads more naturally.
+addWithinRangePred ::
+  RegisterInfo (ArchReg arch) =>
+  BlockIntraStackConstraints arch ids -> NatRepr n -> Integer -> Value arch ids (BVType n) -> Integer ->
+  Either String (BranchConstraints arch ids)
+addWithinRangePred cns w l x u
+  -- Note: we could avoid this duplication by instead producing a conjunction of `addLowerBoundPred`
+  -- and `addUpperBoundPred`, but this produces a single range instead.
+  | l >= maxUnsigned w = Left "Lower bound excludes all possible values"
+  | u < 0 = Left "Upper bound excludes all possible values"
+  | clamp w u < clamp w l = Left "Empty range excludes all possible values"
+  | otherwise =
+    addRangePred cns (intraStackValueExpr cns x) $! mkRangeBound w (fromInteger l) (fromInteger u)
+
+-- | @addExcludeRangePred cns w x l u@ adds a @(x < l) or (u < x)@ constraint over the
+-- `w`-bit-truncated values.  Fails when the bound excludes all possible values.
+addExcludeRangePred ::
+  RegisterInfo (ArchReg arch) =>
+  BlockIntraStackConstraints arch ids -> NatRepr n -> Value arch ids (BVType n) -> Integer -> Integer ->
+  Either String (BranchConstraints arch ids)
+addExcludeRangePred cns w x l u
+  | l <= 0 && u >= maxUnsigned w = Left "Excluded range excludes all possible values"
+  | otherwise =
+    -- compiles it to @(x <= l - 1) or (u + 1 <= x)@
+    disjoinBranchConstraints
+      <$> addUpperBoundPred cns w x (l - 1)
+      <*> addLowerBoundPred cns w (u + 1) x
+
+-- | @assertEqPred cns x w c isTrue@ asserts equality @x = BVValue c w@ is either true or false,
+-- based on `isTrue`.
+assertEqPred ::
+  RegisterInfo (ArchReg arch) =>
+  -- | Constraints that let us evaluate `x`
+  BlockIntraStackConstraints arch ids ->
+  -- | Bitwidth the comparison should apply to
+  NatRepr w ->
+  -- | Value to be constrained
+  Value arch ids (BVType w) ->
+  -- | Numeric constant x should be equal/different to
+  Integer ->
+  -- | `True` if we should assert equality, `False` if we should assert non-equality
+  Bool ->
+  Either String (BranchConstraints arch ids)
+assertEqPred cns w x c isTrue
+  | isTrue =
+    -- x == c becomes c <= x <= c
+    addWithinRangePred cns w c x c
+  | otherwise =
+    -- !(x == c) becomes !(c <= x <= c)
```

Comment:
Perhaps rewrite `!(c <= x <= c)` as something like `(x < c) or (c < x)`, which is more directly what `addExcludeRangePred` encodes.